### PR TITLE
Improve Compat Tests With Python based Compare

### PR DESCRIPTION
**Pull request by Megalonia**
In this PR, we improve our testing from the current tail, head, grep, JQ method and move towards a fixture based approach using python3. The script reads in from std ( nft -j output) and takes a single argument ( fixture ) to compare. The script first checks if they keys do not match or if we should ignore the key, then we check if the value is either a dict/list (recursive cases) or if the values do not match. if all those check pass those keys match and we move onto the next iteration.

to run test's test : `make compat-self-test`

### CODE REVIEW

File: tests/compat-bpf.sh [hunk 1]
```diff
@@ -44,34 +44,11 @@ begin_test "rule should already exist"
 )
 end_test
 
-begin_test "validate nft output: hook"
+begin_test "json diff"
 (
-    $NFT_LIST_TABLE ip $TABLE
-    $NFT_LIST_TABLE_JSON ip $TABLE | $JQ .nftables[].chain.hook | grep input
+    $NFT_LIST_TABLE_JSON ip $TABLE | python3 $BASEDIR/tests/py/compare.py $BASEDIR/tests/fixtures/ip-fixture-filter.json
```
Comment (joewilliams) [inline tests/compat-bpf.sh:49]: Please rename the fixture files to something like `compat-bpf-pinned.json`, `compat-bpf-filter.json` and `compat-sets.json`

**Review by joewilliams — COMMENTED** [review]
We should run `python -m doctest` during docker-ci, before the compat tests. Please add a new section to the make file for that and have it run before any time compat does.

File: tests/compat-bpf.sh [hunk 2]
```diff
@@ -96,28 +73,14 @@ begin_test "create table, chain and bpf rule"
...
 
-begin_test "validate nft output: xt field"
-(
-    $NFT_LIST_TABLE ip $TABLE
-    $NFT_LIST_TABLE_JSON ip $TABLE | $JQ .nftables[].rule.expr | grep xt
-)
-end_test
-
-begin_test "validate nft output: drop verdict"
+begin_test "python compare"
```
Comment (joewilliams) [inline tests/compat-bpf.sh:76]: We called this "json diff" in the other test, lets do that here and elsewhere too.

File: Makefile [hunk 3]
```diff
@@ -12,7 +12,10 @@ input-filter-bpf:
 test:
 	go test $(GOFLAGS) -race -cover -coverprofile=coverage.out ./...
 
-compat-test: input-filter-sets input-filter-bpf
+python-doctest:
```
Comment (joewilliams) [inline Makefile:15]: Rename this to `compat-self-test` please.

**Comment by Megalonia** [comment]
image

metering statments arent being copied over when using `sudo nft -j table foo bar`
I added 'stmt' to be ignored and stopped comparing length for early termination

File: tests/compat-bpf.sh [hunk 4]
```diff
@@ -96,28 +73,14 @@ begin_test "create table, chain and bpf rule"
...
 )
 end_test
-
 begin_test "delete table"
 (
     nft delete table ip $TABLE
 )
 end_test
 
-begin_test "unlink ebpf"
```
Comment (joewilliams) [inline tests/compat-bpf.sh:119]: This test should not be deleted.

**Review by joewilliams — APPROVED** [review]

**Pull request merged**